File: server/src/modules/llm/adapters/openai.adapter.ts
import { LLMAdapter, ChatMessage } from '../llm.interface';

export class OpenAIAdapter implements LLMAdapter {
  constructor(
    private apiKey: string,
    private baseUrl: string = 'https://api.openai.com/v1',
  ) {}

  async *chat(messages: ChatMessage[], stream: boolean): AsyncIterable<string> {
    const response = await fetch(`${this.baseUrl}/chat/completions`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify({
        model: 'gpt-4o-mini',
        messages,
        stream,
      }),
    });

    if (!response.ok) {
      const error = await response.text();
      throw new Error(`OpenAI API error (${response.status}): ${error}`);
    }

    if (!stream) {
      const data = await response.json();
      if (data.error) {
        throw new Error(`OpenAI API error: ${data.error.message}`);
      }
      yield data.choices[0].message.content;
      return;
    }

    const reader = response.body?.getReader();
    if (!reader) {
      throw new Error('Failed to get response stream');
    }

    const decoder = new TextDecoder();
    let buffer = '';

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split('\n');
      buffer = lines.pop() || '';

      for (const line of lines) {
        if (line.startsWith('data: ')) {
          const data = line.slice(6);
          if (data === '[DONE]') return;

          try {
            const parsed = JSON.parse(data);
            if (parsed.error) {
              throw new Error(`OpenAI API error: ${parsed.error.message}`);
            }
            const content = parsed.choices[0]?.delta?.content;
            if (content) yield content;
          } catch (e) {
            if (e instanceof SyntaxError) continue;
            throw e;
          }
        }
      }
    }
  }
}

File: backend/services/llm_gateway/adapters/openai_adapter.py
"""
OpenAI adapter for GPT models.
"""
from typing import Iterator, List, Dict
from openai import OpenAI
from .base import BaseLLMAdapter


class OpenAIAdapter(BaseLLMAdapter):
    """Adapter for OpenAI API."""

    def __init__(self, api_key: str, model: str, base_url: str = None):
        super().__init__(api_key, model, base_url)
        self.client = OpenAI(
            api_key=api_key,
            base_url=base_url if base_url else None
        )

    def stream_chat(self, messages: List[Dict[str, str]]) -> Iterator[str]:
        """Stream chat completion from OpenAI."""
        stream = self.client.chat.completions.create(
            model=self.model,
            messages=messages,
            stream=True
        )

        for chunk in stream:
            if chunk.choices and chunk.choices[0].delta.content:
                yield chunk.choices[0].delta.content

    def chat(self, messages: List[Dict[str, str]]) -> str:
        """Get complete chat response from OpenAI."""
        response = self.client.chat.completions.create(
            model=self.model,
            messages=messages
        )
        return response.choices[0].message.content

    def test_connection(self) -> str:
        """Test OpenAI API connection."""
        response = self.client.chat.completions.create(
            model=self.model,
            messages=[{"role": "user", "content": "Hello"}],
            max_tokens=5
        )
        return f"Connection successful. Model: {self.model}"
